include/openenclave/internal/allocator.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _OE_ALLOCATOR_H
+#define _OE_ALLOCATOR_H
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/bits/types.h>
+
+OE_EXTERNC_BEGIN
+
+/*
+**==============================================================================
+**
+** This header file defines the plugable allocator interface. The default
+** versions of these functions are weakly-typed and included in liboealloc.a.
+** To override these functions, either (1) define strongly-typed functions with
+** the same names, or (2) replace liboealloc.a.
+**
+**==============================================================================
+*/
+
+typedef struct _oe_allocator_stats
+{
+    uint64_t peak_system_bytes;
+    uint64_t system_bytes;
+    uint64_t in_use_bytes;
+} oe_allocator_stats_t;
+
+typedef struct _oe_allocator_upcalls
+{
+    void* (*memset_func)(void* s, int c, size_t n);
+    void* (*memcpy_func)(void* dest, const void* src, size_t n);
+    int (*strcmp_func)(const char* s1, const char* s2);
+    void* (*sbrk_func)(ptrdiff_t increment);
```

Comment:
So I am a little concerned about this being the API that all the allocators will use.  Most modern allocators assume some alignment constraints on the blocks they are using to minimise the cost to find the meta-data. With sbrk this is not available.  

If the allocator is in charge of the whole heap address space, then a layer can be applied to split up the address space into aligned chunks.  However, as OE is using oe_sbrk inside at_exit, then this really forces the use of sbrk by the allocator as it has to correctly interoperate with at_exit. 

I am concerned for any more recent allocator than dlmalloc, you are going to find this can lead to fragmentation in the address space.

I believe any allocator that is capable of being LD_PRELOADed with a libc should be callable by atexit and __cxa_atexit. Most implementation of atexit I have seen just call malloc/calloc after the initial array of 32 elements in exhausted.

Reply:
I guess what I would really like to see here.  Is that the `oe_sbrk` basically become internal to the oealloc, and then the upcalls just expose 
```
__oe_get_heap_base();
__oe_get_heap_end();
```
And it is the responsibility of the allocator to manage the whole heap,

Reply:
I agree that it seems weird to constrain the implementation of the allocator to the subset of the methods defined by `oe_allocator_upcalls_t`, in particular the use of `sbrk`. I think @mjp41's suggestion of letting the allocator implementation become the heap manager makes a lot of sense, and it echoes my personal biases in terms of customizing Windows heap managers.

I'd also like us not to have to arbitrate the rest of the set of "upcalls" as part of this interface, it seems largely based on our current implementation, and the set of helper libc functions could vary from allocator implementation to implementation. In that vein, have we considered simply allowing the replacement to happen at the level of the `oe_malloc` family of symbols?

- malloc.c in its remaining incarnation is only a redirection wrapper between debug and production allocators. Is there a specific value in having OE SDK define the debug interface?
    - Our current debug interface is currently very sparse and not deeply integrated into our logging infrastructure at this point, and it seems unnecessarily restrictive that the debug info is limited to what we define in `oe_allocator_stats_t`.
    - Particularly if the allocator is allowed to manage the entire heap space, I can imagine other metrics being interesting for a custom debug allocator to want to report (e.g. fragmentation, alloc size classes in use).
    - It seems like we could make a debug allocator library a peer of the production one instead of the OE internal defines used to control its use, which is rather less transparent.

- We've not discussed the user expectations for this extension model. Given the internal/allocator.h header, I infer that this falls into the "internal" extension bucket (i.e. you have cloned our sources, so is not usable by consumers of the SDK package).
    - Given that's the case, I would suggest that a developer has full access to the rest of the core libc header surface rather than what we provide through `oe_allocator_upcalls_t` and that they are responsible for managing their circular dependencies on those using it as an example, but not an interface. 
    - For example `oe_allocator_upcalls_t` may only expose `oe_get_heap_base()`/`oe_get_heap_end()` and the dev would be responsible for handling circular dependencies, if any, arising from similar use patterns. For the common ones (memcpy etc.), the chances are good that they have already been pulled in by the linker and won't have to be dealt with specially.

Reply:
> * It seems like we could make a debug allocator library a peer of the production one instead of the OE internal defines used to control its use, which is rather less transparent.

So my concern with this, is that in the code a few key places (pre-this PR), dlmalloc is directly called, such as during unwinding. If the debug allocator is made a peer then this would not be possible.  Maybe that is okay, but I didn't understand the reasoning behind those direct calls, I assumed something went wrong without them.

Reply:
Thank you both. I'll try to address all the points raised above.
- I removed the upcall interface and am allowing the allocator to call anything in core for now.
- atexit() now depends on malloc() rather than sbrk(), so the allocator can now use the heap boundaries if it likes. But beware:
    - Third-parties libraries are free to use sbrk().
    - Most allocators I've used coexist with sbrk().
    - Be sure your allocator does not have a circular dependency on malloc.
- Moot now, but calling sbrk(NULL) gives you the current break value, and you can use this to determine your alignment requirements.
- I've left the debug allocator as is. I don't want to lose the capabilities when using another allocator.

Reply:
> Moot now, but calling sbrk(NULL) gives you the current break value, and you can use this to determine your alignment requirements.

This is not thread safe though, as another call can come in between.

> Third-parties libraries are free to use sbrk().

Do you think many will?  Is this something you really want to support?  I did a very unscientific search of GitHub for code using sbrk, and it seemed primarily to be CS homework assignments for writing allocators. 

The man page for sbrk says it should be avoided and was removed from POSIX.2001-1.